x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -834,7 +837,23 @@ protected void doSaveState(IndexerState indexerState, DataFrameIndexerPosition p
                                     onStop();
                                     transformTask.shutdown();
                                 }
-                                next.run();
+                                // Only do this clean up once, if it succeeded, no reason to do the query again.
+                                if (oldStatsCleanedUp.compareAndExchange(false, true) == false) {
```

Comment:
nit: compareAndSet is cleaner as you don't need the negation.

`if (oldStatsCleanedUp.compareAndSet(false, true) == false)`